sim-se: zero out memory allocated via brk()

nmosier opened the pull request:
The syscall emulation of brk() incorrectly did not ensure that newly allocated memory was zero-initialized, which Linux guarantees and which seems to be the expectation of glibc's malloc() and free() implementation. This patch fixes the incorrect behavior by zero- initalizing all memory allocations via brk().

GitHub issue: https://github.com/gem5/gem5/issues/342

Change-Id: I53cf29d6f3f83285c8e813e18c06c2e9a69d7cc2

giactra commented on src/sim/syscall_emul.cc:
```diff
@@ -270,6 +270,13 @@ brkFunc(SyscallDesc *desc, ThreadContext *tc, VPtr<> new_brk)
 
     mem_state->updateBrkRegion(brk_point, new_brk);
 
+    if (new_brk > brk_point) {
```
Nit: Could you add a comment here just describing the zeroing?

giactra reviewed (changes requested):
That's interesting. I looked at the brk(2) man pages [1][2] and it does not reference any sort of zeroing happening when the program break gets increased. I know this is automatically done when the allocator ask the OS memory with mmap+/dev/zero but I don't see any zeroing being referenced by the linux kernel with the program break.

Could you point me to some documentation describing this behaviour?

[1]: https://man7.org/linux/man-pages/man2/brk.2.html
[2]: https://linux.die.net/man/2/brk

nmosier committed:
sim-se: unmap reclaimed heap pages in brk syscall emulation

gem5::MemState::updateBrkRegion(), which is called during the syscall
emulation of brk, did not unmap deallocated heap pages when the brk
region is receding. Instead, it kept it mapped for simplicity. This
introduced a bug where subequent expansions of the brk region reused
prior heap page mappings that were not zero-filled. This violates
the assumptions of glibc malloc, resulting in heap corruption and
crashes.

This patch fixes the bug by always unmapping pages that are deallocated
during a call to brk() that reduces the heap size. This makes the
gem5::MemState::_endBrkPoint field obsolete, so this patch removes it.

GitHub issue: https://github.com/gem5/gem5/issues/342

Change-Id: Ib2244e1aa4d2a26666ad60d231fdde2c22d2df35

nmosier commented:
I look into the issue a bit further, and it looks like the real bug is due to the fact that the `gem5::MemState::updateBrkRegion()` method does not unmap deallocated heap pages when the program break recedes. While I can't point to documentation yet showing that pages newly allocated with brk() should be zero-filled, I was able to find exactly where in glibc malloc's source that they encode this assumption: [malloc/malloc.c, line 517](https://sourceware.org/git?p=glibc.git;a=blob;f=malloc/malloc.c;h=d0bbbf371048ee8aa8a30c03b189cb268b8ad9e4;hb=refs/heads/master#l517):
```c
/*  MORECORE_CLEARS           (default 1)                                                                                                                                                                                                     
     The degree to which the routine mapped to MORECORE zeroes out                                                                                                                                                                            
     memory: never (0), only for newly allocated space (1) or always                                                                                                                                                                          
     (2).  The distinction between (1) and (2) is necessary because on                                                                                                                                                                        
     some systems, if the application first decrements and then                                                                                                                                                                               
     increments the break value, the contents of the reallocated space                                                                                                                                                                        
     are unspecified.                                                                                                                                                                                                                         
 */
```
and in [sysdeps/unix/sysv/linux/Makefile](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/Makefile;h=063719bae645b24a4474729640453fa4b5a63056;hb=refs/heads/master) they set the default for Linux to be 2, i.e., memory returned by MORECORE (which defaults to sbrk()) is always zero-filled:
```
ifeq ($(subdir),malloc)
CFLAGS-malloc.c += -DMORECORE_CLEARS=2
endif
```

What I had before, zero-filling newly allocated brk() memory directly, would also work, but the Linux kernel doesn't appear to do that. It seems to only unmap freed pages when the program break recedes.

nmosier commented:
I'm wondering if the test that failed is flaky, since all tests passed locally on my machine.

powerjg commented:
Yeah, we're having intermittent network issues with some of our runners where the test binaries are failing to download. Hopefully it will work when we re-run it.

giactra commented:
> I look into the issue a bit further, and it looks like the real bug is due to the fact that the `gem5::MemState::updateBrkRegion()` method does not unmap deallocated heap pages when the program break recedes. While I can't point to documentation yet showing that pages newly allocated with brk() should be zero-filled, I was able to find exactly where in glibc malloc's source that they encode this assumption: [malloc/malloc.c, line 517](https://sourceware.org/git?p=glibc.git;a=blob;f=malloc/malloc.c;h=d0bbbf371048ee8aa8a30c03b189cb268b8ad9e4;hb=refs/heads/master#l517):
> 
> ```c
> /*  MORECORE_CLEARS           (default 1)                                                                                                                                                                                                     
>      The degree to which the routine mapped to MORECORE zeroes out                                                                                                                                                                            
>      memory: never (0), only for newly allocated space (1) or always                                                                                                                                                                          
>      (2).  The distinction between (1) and (2) is necessary because on                                                                                                                                                                        
>      some systems, if the application first decrements and then                                                                                                                                                                               
>      increments the break value, the contents of the reallocated space                                                                                                                                                                        
>      are unspecified.                                                                                                                                                                                                                         
>  */
> ```
> 
> and in [sysdeps/unix/sysv/linux/Makefile](https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/Makefile;h=063719bae645b24a4474729640453fa4b5a63056;hb=refs/heads/master) they set the default for Linux to be 2, i.e., memory returned by MORECORE (which defaults to sbrk()) is always zero-filled:
> 
> ```
> ifeq ($(subdir),malloc)
> CFLAGS-malloc.c += -DMORECORE_CLEARS=2
> endif
> ```
> 
> What I had before, zero-filling newly allocated brk() memory directly, would also work, but the Linux kernel doesn't appear to do that. It seems to only unmap freed pages when the program break recedes.

Thanks @nmosier for taking the time to look into the documentation. Pointing to the glibc implementation really helps on clearing out some doubts!

giactra commented on src/sim/mem_state.cc:
```diff
@@ -107,26 +106,34 @@ MemState::isUnmapped(Addr start_addr, Addr length)
+     * the * old_brk; this occurs when the heap is receding because the
+     * application has given back memory. This may involve unmapping
+     * heap pages, if new_brk rounds to a lower-address page. The
+     * previous, buggy behavior was to leave such pages mapped for
+     * simplicity. However, this introduced a bug where, if the heap
+     * was later expanded out, this routine reuses previously mapped
+     * heap pages that were not zero-filled (and instead were filled
+     * with stale application data), violating the assumptions of
+     * glibc malloc.
```
Nit: I would reword this part of the comment as IMHO it's not entirely accurate to label previous behaviour as intrinsically buggy.

From what I understood not zeroing out the old heap is technically legal and it does not violate any glibc assumption as long as glibc is informed by such behaviour via the MORECORE_CLEAR value. We should just note that Linux expect the zeroing and unmapping pages anyway regardless of the MORECORE_CLEAR is the safest option that works for every system.

giactra reviewed (changes requested):
Code looks great I just have a minor comment

powerjg reviewed (approved):
lgtm! Thanks for the fix
